### PR TITLE
Correct inference of primitive operand type behind binary operation

### LINKED ISSUE #57447: Binary operations involving &1.0 always infer f64
Originally posted as a comment on #36549, but in hindsight this may be a different issue. (in that `#36549` *very clearly looks like a bug*, whereas this one is in a potential gray area)

[Relevant URLO thread](https://users.rust-lang.org/t/the-iterator-binds-its-associated-type-not-by-the-first-call/23991).

```rust
let _: f32 = 1. - 1.;   // allowed
let _: f32 = 1. - &1.;  // type error
let _: f32 = &1. - 1.;  // type error
let _: f32 = &1. - &1.; // type error
```

In the latter 3 cases, the type of `&1.` is apparently defaulted to `&f64`:

```
   Compiling playground v0.0.1 (/playground)
error[E0271]: type mismatch resolving `<f64 as std::ops::Sub<&f64>>::Output == f32`
 --> src/main.rs:3:21
  |
3 |     let _: f32 = 1. - &1.; // type error
  |                     ^ expected f64, found f32

error[E0271]: type mismatch resolving `<&f64 as std::ops::Sub<f64>>::Output == f32`
 --> src/main.rs:4:22
  |
4 |     let _: f32 = &1. - 1.; // type error
  |                      ^ expected f64, found f32

error[E0271]: type mismatch resolving `<&f64 as std::ops::Sub<&f64>>::Output == f32`
 --> src/main.rs:5:22
  |
5 |     let _: f32 = &1. - &1.; // type error
  |                      ^ expected f64, found f32
```

To a mortal like me, it seems that the only reason the first line works is because the compiler must have a special case for binary operations between two unconstrained "floating-point flavored" type inference variables. Can the compiler not just special case the latter three examples in the same way it special cases the first?

### PR DESCRIPTION
**Pull request by varkor**
Fixes https://github.com/rust-lang/rust/issues/57447.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_typeck/check/op.rs [hunk 1]
```diff
@@ -864,6 +873,14 @@ enum Op {
     Unary(hir::UnOp, Span),
 }
 
+/// Dereferences a single level of immutable referencing.
+fn deref_ty<'tcx>(ty: Ty<'tcx>) -> Ty<'tcx> {
```
Comment (nagisa) [inline src/librustc_typeck/check/op.rs]: this function needs a better name.

**Comment by nikomatsakis** [comment]
The code seems fine, but I think we probably ought to document the behavior of this part of the type-checker better. This effects the lang spec in the end. Perhaps the reference is the right place? I don't think we have a lot of "framework" for doing so, but I think it's probably a good idea for us to start collecting these sorts of things in the reference.

This is tricky, though, as we don't have any sort of "framework", and it's hard to describe the structure of the rules without that. Still, I wouldn't want to get too hung up on (e.g) figuring out the best way to write typing judgements.

Maybe we should start by documenting in rustc-guide or in an open issue to start?

**Comment by varkor** [comment]
I've written a [comment on a relevant issue](https://github.com/rust-lang/reference/issues/104#issuecomment-574356163) in the Rust reference repository.

**Comment by varkor** [comment]
I find it slightly surprising that `x + y` and `&x + &y` work, but `&&x + &&y` (and similar) do not. This is due to https://github.com/rust-lang/rust/blob/056d48a1bc454dec2186f9242db77af69e493524/src/libcore/internal_macros.rs#L21-L23
which seems a little ad hoc (added in https://github.com/rust-lang/rust/pull/21227).

**Comment by nikomatsakis** [comment]
It seems a bit surprising, but then those are the impls that exist, right? (i.e., we don't have a general impl for `&T`, just for specific types?)

**Comment by varkor** [comment]
Yes, that's right. It's related, but only tangentially in that the cases look similar from a user's perspective.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

Discussed in [today's @rust-lang/lang meeting](https://github.com/rust-lang/lang-team/blob/master/minutes/2020-01-16.md). The consensus of the meeting was that we should make this change, but we're doing an fcp merge to make it official.

We did think it would be good to add some more tests:

* To cover other sorts of operators, like `<<`
* To test cases where the integer is "indirect", [like this one](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=431b1d87d4a5e8a895b86d3f365cbc43)
* To test cases where there is no expected type (which we expect to error)

**Comment by nikomatsakis** [comment]
@bors r-

Wait, @varkor, did you add the extra tests we mentioned in the meeting...?

**Comment by varkor** [comment]
@nikomatsakis: no, I was waiting for the FCP to complete. I'll do so soon.

**Comment by varkor** [comment]
> To test cases where there is no expected type (which we expect to error)

Just to clarify, do you mean cases like the following?
```rust
let _ = 0 + &0;
```
This currently passes, as it defaults to `i32`. Or was a different case supposed to error?

I've added the suggested test cases.

**Comment by nikomatsakis** [comment]
Good question @varkor, what *did* I mean... I'm not sure, but the tests seem good.

@bors r+

**Comment by MightyPork** [comment]
I hoped this would fix the ever-present "you need a *" nag, but e.g. this still does not work - is that split to a separate issue? It seems to still be broken for all comparison operators

```
fn main() {
    let a = 1u16;
    let b = 0u16;
    
    let a_borrowed = &a;
    
    if a_borrowed > b { // expected `&u16`, found `u16`
        println!("hello");
    }
}
```

**Comment by varkor** [comment]
This would be a separate issue, but I'm not sure whether or not it's covered by an existing issue.
The reason this doesn't currently work is that we don't call`forward_ref_binop` https://github.com/rust-lang/rust/blob/a59264b01247836c70e24217e0d346b868387525/src/libcore/internal_macros.rs#L21-L23
for the comparison operators, unlike the arithmetic operators. I'm not sure if this is intentional or not. You're welcome to open an issue: if this discussion already exists, someone else will know about it.

**Pull request merged**